os: expose `guessFileDescriptorType`

vladfrangu opened the pull request:
Exposes the internal `guessHandleType` function, which can be used to see if a handle has a specific type, regardless of the OS it is on.

This helps out with detecting, for example, if standard input is piped into the process, instead of relying on file system calls.

Refs: https://github.com/nodejs/node/issues/57603

vladfrangu committed:
os: expose `guessFileDescriptorType`

Exposes the internal `guessHandleType` function as
`guessFileDescriptorType`, which can be used to see if a handle has a
specific type, regardless of the OS it is on.

This helps out with detecting, for example, if standard input is piped
into the process, instead of relying on file system calls.

Refs: https://github.com/nodejs/node/issues/57603

vladfrangu committed:
fix: validate fd in JS land + add test

vladfrangu committed:
chore: fix test and add more checks

aduh95 commented on doc/api/os.md:
```diff
@@ -509,6 +509,48 @@ On POSIX systems, the operating system release is determined by calling
+
+Currently, the following types for a handle can be returned:
+
+<table>
+  <tr>
+    <th>Constant</th>
+  </tr>
+  <tr>
+    <td><code>TCP</code></td>
+  </tr>
+  <tr>
+    <td><code>TTY</code></td>
+  </tr>
+  <tr>
+    <td><code>UDP</code></td>
+  </tr>
+  <tr>
+    <td><code>FILE</code></td>
+  </tr>
+  <tr>
+    <td><code>PIPE</code></td>
+  </tr>
+  <tr>
+    <td><code>UNKNOWN</code></td>
+  </tr>
+  <tr>
+    <td><code>INVALID</code></td>
+  </tr>
+</table>
```
Why not use markdown table or list? That would be much more readable

vladfrangu replied:
I honestly debated adding a secondary column describing the meaning of each of the values, but then I realized that the values are pretty self-explanatory. Will turn to a normal markdown list 👍

vladfrangu replied:
Pushed, not sure if this is what you were referring to (also not sure if I should wrap the values in quotes or not)

aduh95 commented on lib/internal/util.js:
```diff
@@ -864,7 +865,13 @@ function getCIDR(address, netmask, family) {
 }
 
 const handleTypes = ['TCP', 'TTY', 'UDP', 'FILE', 'PIPE', 'UNKNOWN'];
+handleTypes[-1] = 'INVALID';
```
```suggestion
setOwnProperty(handleTypes, -1, 'INVALID');
```

vladfrangu committed:
chore: suggested changes

vladfrangu commented:
I'll rebase to fix the first commit message after the CI passes! Should I also squash the PR or leave the rest as is?

aduh95 commented:
> Should I also squash the PR or leave the rest as is?

As you prefer, in any case it will get squashed upon merging.

addaleax commented on doc/api/os.md:
```diff
@@ -509,6 +509,29 @@ On POSIX systems, the operating system release is determined by calling
+added: REPLACEME
+-->
+
+* `handle` {integer} The handle number to try and guess the type of.
```
This should really specify that "handle" means "file descriptor". In fact, given that the fs docs almost exclusively use "file descriptor" and "fd", I would also adopt that terminology here.

vladfrangu replied:
Then I could also rename the function to `guessFileDescriptorType`? Or should I leave it as is?

And will apply the changes to fd/file descriptor in a few

addaleax replied:
Yeah, I'd probably do that. In the context of libuv, `guessHandleType` makes more sense, because it returns a value that actually corresponds to the types of libuv handles, which wrap around fds.

vladfrangu replied:
Done! Let me know if I've also done the rebase correctly 🤞

vladfrangu committed:
chore: rename to os.guessFileDescriptorType

jasnell commented on doc/api/os.md:
```diff
@@ -508,6 +508,29 @@ On POSIX systems, the operating system release is determined by calling
+* Returns: {string}
+
+Returns the type of the file descriptor passed in, or `'INVALID'` if the provided file descriptor
+is invalid.
```
returning `null` instead of `INVALID` might be more idiomatic.

jasnell commented on lib/internal/util.js:
```diff
@@ -865,7 +866,13 @@ function getCIDR(address, netmask, family) {
+setOwnProperty(handleTypes, -1, 'INVALID');
+
 function guessHandleType(fd) {
+  if (!NumberIsInteger(fd)) {
```
In fs APIs we will throw an error if the `fd` is the incorrect type. We probably should do so here for consistency

jasnell commented on lib/os.js:
```diff
@@ -39,7 +39,7 @@ const {
   hideStackFrames,
 } = require('internal/errors');
-const { getCIDR } = require('internal/util');
+const { getCIDR, guessHandleType: _guessHandleType } = require('internal/util');
```
```suggestion
const { getCIDR, guessHandleType: guessFileDescriptorType } = require('internal/util');
```

jasnell commented on lib/os.js:
```diff
@@ -328,6 +328,7 @@ module.exports = {
   uptime: getUptime,
   version: getOSVersion,
   machine: getMachine,
+  guessFileDescriptorType: _guessHandleType,
```
```suggestion
  guessFileDescriptorType,
```

jasnell commented on src/node_util.cc:
```diff
@@ -116,19 +113,15 @@ static void GetPromiseDetails(const FunctionCallbackInfo<Value>& args) {
-      proxy->GetTarget(),
-      proxy->GetHandler()
-    };
+    Local<Value> ret[] = {proxy->GetTarget(), proxy->GetHandler()};
```
There are a large number of unrelated changes in here. Were these manually edited or did the format tool do these?

vladfrangu replied:
All of the formatting changes were done by saving the file (after configuring vscode with the settings example provided by the contributing guide). Not sure exactly _why_ so many changes happened, but running the cpp formatter through make also didn't revert it. I can manually revert it if desired

vladfrangu replied:
So, I did revert these changes, and staged them, but after running `make format-cpp`, the file was reformatted to how it is on this PR. I'll push the non-formatted change but not sure if I should open a PR after formatting the file or just leave it be

vladfrangu replied:
Yet running `CLANG_FORMAT_START=main make format-cpp` changes nothing, confusing.. I'll leave that be for now

vladfrangu committed:
chore: suggested changes

Co-authored-by: James M Snell <jasnell@gmail.com>

vladfrangu committed:
chore: requested changes

vladfrangu committed:
chore: remove wrong primordial destructure

aduh95 commented on test/pseudo-tty/test-os-guessFileDescriptorType.js:
```diff
@@ -0,0 +1,19 @@
+strictEqual(guessFileDescriptorType(55555), 'UNKNOWN', '55555 reported to be a handle, but it is not');
+strictEqual(guessFileDescriptorType(2 ** 31 - 1), 'UNKNOWN', '2^31-1 reported to be a handle, but it is not');
+
+throws(() => guessFileDescriptorType(-1), /"fd" must be a positive integer/, '-1 reported to be a handle, but it is not');
+throws(() => guessFileDescriptorType(1.1), /"fd" must be a positive integer/, '1.1 reported to be a handle, but it is not');
+throws(() => guessFileDescriptorType('1'), /"fd" must be a positive integer/, '\'1\' reported to be a tty, but it is not');
+throws(() => guessFileDescriptorType({}), /"fd" must be a positive integer/, '{} reported to be a tty, but it is not');
+throws(() => guessFileDescriptorType(() => {}), /"fd" must be a positive integer/, '() => {} reported to be a tty, but it is not');
+throws(() => guessFileDescriptorType(2 ** 31), /"fd" must be a positive integer/, '2^31 reported to be a handle, but it is not (because the fd check rolls over the input to negative of it)');
```
We should be checking for a error code, not the message (error messages are not covered by semver, error codes are)
```suggestion
[
    -1,
    1.1,
    '1',
    [],
    {},
    () => {},
    2 ** 31,
    true,
    false,
    1n,
    Symbol(),
    undefined,
    null,
].forEach((val) => throws(() => guessFileDescriptorType(val), { code: 'ERR_INVALID_FD' }));
```

vladfrangu committed:
chore: use forEach and assert based on error code

Co-authored-by: Antoine du Hamel <duhamelantoine1995@gmail.com>

vladfrangu committed:
chore: remove reformatting changes

vladfrangu committed:
chore: lint js file

vladfrangu committed:
fix: test fail

jasnell commented on doc/api/os.md:
```diff
@@ -508,6 +508,28 @@ On POSIX systems, the operating system release is determined by calling
+* Returns: {string|null}
+
+Returns the type of the file descriptor passed in, or `null` if the provided file descriptor
+is invalid.
```
Nit: some expanded explanation about why this is useful would probably be helpful. It's rather obscure.

jasnell commented on lib/internal/util.js:
```diff
@@ -876,7 +877,13 @@ function getCIDR(address, netmask, family) {
 }
 
 const handleTypes = ['TCP', 'TTY', 'UDP', 'FILE', 'PIPE', 'UNKNOWN'];
+setOwnProperty(handleTypes, -1, null);
```
this feels almost a bit too clever. Why not simply have `_guessHandleType(fd)` return null explicitly? then your return value below could be `return handletypes[type] || type`.

vladfrangu commented on src/node_util.cc:
```diff
@@ -216,15 +216,24 @@ static uint32_t GetUVHandleTypeCode(const uv_handle_type type) {
+      // For an unhandled handle type, we want to return `UNKNOWN` instead of
+      // `null` since the type is "known" by UV, just not exposed further to
+      // JS land
+      return 5;
```
I would do it if I knew how 😅, and there is a TODO for that already in the function

vladfrangu committed:
chore: simplify return type

Tested with `require('internal/test/binding').internalBinding('util').guessHandleType(2**31)` (not sure if there was a better way, but it works so)

vladfrangu committed:
docs: try to explain the use case for this function

vladfrangu committed:
chore: fix lint

Signed-off-by: Vlad Frangu <me@vladfrangu.dev>

vladfrangu commented:
Rebased again, hoping CI also passes now :D

vladfrangu commented:
PR rebased, hoping CI works now? 😅

vladfrangu commented:
Should I rebase the PR again? Also fixing the commit message with the missing signed-off?

aduh95 commented:
> Should I rebase the PR again? Also fixing the commit message with the missing signed-off?

It's the only way to get green CI, so I'd say yes

vladfrangu commented:
Should be correct and done now, let me know please 🙏
